conjure-codegen/src/context.rs

```diff
@@ -1002,6 +1028,84 @@ impl Context {
         }
     }
 
+    fn type_log_safety(&self, ty: &Type) -> Option<LogSafety> {
+        match ty {
+            Type::Primitive(primitive) => self.primitive_log_safety(primitive),
+            Type::Optional(optional) => self.type_log_safety(optional.item_type()),
+            Type::List(list) => self.type_log_safety(list.item_type()),
+            Type::Set(set) => self.type_log_safety(set.item_type()),
+            Type::Map(map) => self.combine_safety(
+                self.type_log_safety(map.key_type()),
+                self.type_log_safety(map.value_type()),
+            ),
+            Type::Reference(def) => self.type_log_safety_ref(def),
+            Type::External(_) => None,
+        }
+    }
+
+    fn primitive_log_safety(&self, primitive: &PrimitiveType) -> Option<LogSafety> {
+        match primitive {
+            PrimitiveType::Bearertoken => Some(LogSafety::DoNotLog),
+            _ => None,
+        }
+    }
+
+    fn type_log_safety_ref(&self, name: &TypeName) -> Option<LogSafety> {
+        let ctx = &self.types[name];
+
+        if let CachedLogSafety::Computed(safety) = &*ctx.log_safety.borrow() {
+            return safety.clone();
+        }
+
+        // temporarily treat it as safe in case of recursive type definitions.
+        *ctx.log_safety.borrow_mut() = CachedLogSafety::Computed(Some(LogSafety::Safe));
+
+        let safety = match &ctx.def {
+            TypeDefinition::Alias(alias) => alias
+                .safety()
+                .cloned()
+                .or_else(|| self.type_log_safety(alias.alias())),
+            // We consider enums to be safe even when not compiled as exhaustive, since we assume
+            // unknown variants are simply from a future definition.
+            TypeDefinition::Enum(_) => Some(LogSafety::Safe),
+            TypeDefinition::Object(object) => object
+                .fields()
+                .iter()
+                .map(|f| {
+                    f.safety()
+                        .cloned()
+                        .or_else(|| self.type_log_safety(f.type_()))
+                })
+                .fold(Some(LogSafety::Safe), |a, b| self.combine_safety(a, b)),
+            TypeDefinition::Union(union_) => union_
+                .union_()
+                .iter()
+                .map(|f| {
+                    f.safety()
+                        .cloned()
+                        .or_else(|| self.type_log_safety(f.type_()))
+                })
+                // The unknown variant is unsafe to log, and we don't want log safety to vary based
+                // on the type generation configuration.
+                .fold(Some(LogSafety::Unsafe), |a, b| self.combine_safety(a, b)),
```

Comment:
Just acknowledging that this is a different decision than the Java implementation.

Reply:
Ah - I misread the Java logic! I will update to start with None.

Reply:
It was actually changed after the initial implementation: https://github.com/palantir/conjure-java/commit/4b08523c3b62455544f7b0544f07ae74eee4a13c